wasm/src/object.c

```diff
@@ -435,10 +435,47 @@ opa_value *builtin_object_get(opa_value *obj, opa_value *key, opa_value *value)
         return NULL;
     }
 
-    opa_object_elem_t *elem = opa_object_get(opa_cast_object(obj), key);
-    if (elem != NULL)
+    // if the key is not an array, then we get that top level key from the object or return the default value
+    if (opa_value_type(key) != OPA_ARRAY) {
+        opa_object_elem_t *elem = opa_object_get(opa_cast_object(obj), key);
+        if (elem != NULL)
+        {
+            return elem->v;
+        }
+
+        return value;
+    }
+
+    size_t path_len = opa_cast_array(key)->len;
```

Comment:
https://github.com/open-policy-agent/opa/blob/b186719e849283a82c559f9f5ea519c92d87ee5b/wasm/src/value.c#L1496-L1498 Maybe `_validate_json_path` could be of use here? (Genuine question, it's been a while since I've touched this 😅)

Reply:
Hmm, I thought this might work but get the following error locally in https://github.com/open-policy-agent/opa/pull/4166/commits/9a50f7bee7b6737ea99c98deda51ea787451ffb9.

```
charlie_egan@cloudshell:~/opa$ ./opa_linux_amd64 eval -fpretty -t wasm 'object.get({"a": 1}, [], 2)'
panic: runtime error: slice bounds out of range [-1:] [recovered]
        panic: runtime error: slice bounds out of range [-1:]

goroutine 1 [running]:
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel.func2.1()
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:739 +0x24c
panic({0x1a35de0, 0xc000263dd0})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/bytecodealliance/wasmtime-go.enterWasm({0x1be58c0, 0xc0000599c0}, 0xc0004bebc8)
        /home/charlie_egan/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:532 +0xef
github.com/bytecodealliance/wasmtime-go.(*Func).Call(0xc000336e20, {0x1be58c0, 0xc0000599c0}, {0xc0004ae0e0, 0x7, 0x7})
        /home/charlie_egan/opa/vendor/github.com/bytecodealliance/wasmtime-go/func.go:414 +0x20c
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel.func2(0x1d44f, 0x1be58c0, 0xc0000599c0, {0xc0004ae0e0, 0xc0004bed60, 0x0})
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:743 +0xac
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.callOrCancel({0x1c00b10, 0xc000128000}, 0xc0000ac500, {0x1a9d7b4, 0x8}, {0xc0004bedfc, 0x7, 0x0})
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:745 +0x207
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.call({0x1c00b10, 0xc000128000}, 0xc0004bede0, {0x1a9d7b4, 0x68ced7}, {0xc0004bedfc, 0x9b0094, 0x1c00d08})
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:686 +0x2d
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.newVM.func6({0x1c00b10, 0xc000128000}, 0x11, 0x0, 0x3df4b0, 0xc0, 0x599f0)
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:138 +0x73
github.com/open-policy-agent/opa/internal/wasm/sdk/internal/wasm.(*VM).Eval(0xc0000ac500, {0x1c00b10, 0xc000128000}, 0x5e0906, 0x0, {0x1c0dd10, 0xc0003ddbc0}, {0x0, 0x0}, {0x0, ...}, ...)
        /home/charlie_egan/opa/internal/wasm/sdk/internal/wasm/vm.go:335 +0x623
github.com/open-policy-agent/opa/internal/wasm/sdk/opa.(*OPA).Eval(0xc0004ae000, {0x1c00b10, 0xc000128000}, {0x0, 0x0, {0x1c0dd10, 0xc0003ddbc0}, {0x0, 0x0, 0x0}, ...})
        /home/charlie_egan/opa/internal/wasm/sdk/opa/opa.go:193 +0x258
github.com/open-policy-agent/opa/features/wasm.(*OPA).Eval(0xc0003e2780, {0x1c00b10, 0xc000128000}, {0x0, {0x1c0dd10, 0xc0003ddbc0}, 0x0, {0x0, 0x0, 0x0}, ...})
        /home/charlie_egan/opa/features/wasm/wasm.go:71 +0x15c
github.com/open-policy-agent/opa/rego.(*Rego).evalWasm(0xc0001ba300, {0x1c00b10, 0xc000128000}, 0xc00047c2c0)
        /home/charlie_egan/opa/rego/rego.go:1967 +0x1d8
github.com/open-policy-agent/opa/rego.(*Rego).eval(0xc0001ba300, {0x1c00b10, 0xc000128000}, 0xc000128000)
        /home/charlie_egan/opa/rego/rego.go:1894 +0x375
github.com/open-policy-agent/opa/rego.PreparedEvalQuery.Eval({{0xc0001ba300, 0xc0003df630}}, {0x1c00b10, 0xc000128000}, {0xc0003df480, 0x203000, 0x203000})
        /home/charlie_egan/opa/rego/rego.go:393 +0x166
github.com/open-policy-agent/opa/cmd.evalOnce({0x1c00b10, 0xc000128000}, 0xc0003ee000)
        /home/charlie_egan/opa/cmd/eval.go:401 +0x286
github.com/open-policy-agent/opa/cmd.eval({_, _, _}, {0xc0003a65a0, 0x0, 0x0, {0xc000347460, 0x1, 0x1}, {0x26b52f0, ...}, ...}, ...)
        /home/charlie_egan/opa/cmd/eval.go:326 +0x3ea
github.com/open-policy-agent/opa/cmd.init.4.func2(0xc00039aa00, {0xc0003ca140, 0x4, 0x4})
        /home/charlie_egan/opa/cmd/eval.go:251 +0x78
github.com/spf13/cobra.(*Command).execute(0xc00039aa00, {0xc0003ca100, 0x4, 0x4})
        /home/charlie_egan/opa/vendor/github.com/spf13/cobra/command.go:860 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0x25e1940)
        /home/charlie_egan/opa/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
        /home/charlie_egan/opa/vendor/github.com/spf13/cobra/command.go:902
main.main()
        /home/charlie_egan/opa/main.go:15 +0x25
```

I've opted to revert this for now, but with some pointers I can dig a little deeper.